Feature/jet fcst only

lgannoaa opened the pull request:
Support JET GFS fcst only run. Improvement made:
JET environment support
  env/JET.env
JET partition support
  parm/config/config.fv3
  parm/config/config.resources
JET python 3 gfs fcst only scripts
  setup_expt_fcstonly.py, setup_workflow_fcstonly.py, workflow_utils.py
Module
  module_base.jet
Build scripts and exception handling for the script not supported on JET
  build_fv3.sh, build_gfs_util.sh, build_gldas.sh, build_ww3prepost.sh, build_workflow_utils.sh
Link script
  sorc/link_fv3gfs.sh

lgannoaa committed:
JET update to enable fcst only run:
 Changes to be committed:
	modified:   env/JET.env
	modified:   modulefiles/module_base.jet
	modified:   sorc/build_gfs_util.sh
	modified:   sorc/build_gldas.sh
	modified:   sorc/build_workflow_utils.sh
	modified:   sorc/link_fv3gfs.sh

lgannoaa committed:
Include sdditional JET modulefiles and script

lgannoaa committed:
Update hpc-stack and clean up the build_workflow_utils.sh script
Changes to be committed:
	modified:   modulefiles/module_base.jet
	modified:   sorc/build_gldas.sh
	modified:   sorc/build_workflow_utils.sh

lgannoaa committed:
Modified to remove OznMon RadMon modulefiles
Disable JET build for unsupported package gfs_util, gldas, and ww3
Update python scripts in ush/rocoto
	deleted:    modulefiles/OznMonBuild.jet
	deleted:    modulefiles/RadMonBuild.jet
	modified:   sorc/build_fv3.sh
	modified:   sorc/build_gfs_util.sh
	modified:   sorc/build_gldas.sh
	modified:   sorc/build_ww3prepost.sh
	modified:   ush/rocoto/rocoto.py
	modified:   ush/rocoto/setup_expt_fcstonly.py
	modified:   ush/rocoto/setup_workflow_fcstonly.py
	modified:   ush/rocoto/workflow_utils.py
	deleted:    util/sorc/compile_gfs_util_jet.sh

lgannoaa committed:
Modified to use netcdfp/4.7.4 from hfv3gfs/gwv location in module_base.jet

lgannoaa committed:
Remove unused module from module_base.jet

lgannoaa committed:
Remove gfs_util.jet because it is not currently supported on JET

lgannoaa committed:
merge JET python scripts for GFS fcst only workflow with python3 syntax
Changes to be committed:
	modified:   setup_expt_fcstonly.py
	modified:   setup_workflow_fcstonly.py
	modified:   workflow_utils.py

lgannoaa committed:
Now include additional JEt PARTITION support with:
xjet: npe_node_max=24
vjet, sjet: npe_node_max=16
kjet: npe_node_max=40
Changes to be committed:
	modified:   parm/config/config.fv3
	modified:   parm/config/config.resources

lgannoaa committed:
Replace rocoto.py with python3 version

lgannoaa commented:
Testing record:
[ =141 15:44:55 4969 Lin.Gan@fe3 ]
.../HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-develop> cd /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/ush/rocoto/
[ =0 15:45:52 4970 Lin.Gan@fe3 ]
.../HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr> module list
Currently Loaded Modules:
  1) emc-utils/1.2.0   4) intelpython/3.6.5   7) hpc-intel/18.0.5.274  10) hpc-impi/2018.4.274  13) nco/4.9.1     16) prod_util/v1.0.18  19) crtm/2.3.0   22) netcdf/4.7.4
  2) xxdiff/3.2.Z1     5) hpc/1.1.0           8) szip/2.1.1            11) wgrib2/2.0.8         14) cdo/1.9.5     17) grib_util/v1.1.1   20) esmf/8_1_1   23) rocoto/1.3.3
  3) wgrib/1.8.1.0b    6) intel/18.0.5.274    9) impi/2018.4.274       12) hpss                 15) gempak/7.4.2  18) g2tmpl/1.6.0       21) hdf5/1.10.6
[ =0 15:48:28 4976 Lin.Gan@fe3 ]
.../Lin.Gan/PR/global-workflow-pr/ush/rocoto> ./setup_expt_fcstonly_jet.sh
++ USER=Lin.Gan
++ HOMEgfs=/mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr
++ EXPDIR=/lfs4/HFIP/hfv3gfs/Lin.Gan/expdir
++ PSLOT=p3_c192_pr
++ COMROT=/lfs4/HFIP/hfv3gfs/Lin.Gan/ptmp/com
++ IDATE=2020090600
++ EDATE=2020090612
++ RESDET=192
++ ln -fs /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/parm/config/config.base.emc.dyn /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/parm/config/config.base
++ ./setup_expt_fcstonly.py --pslot p3_c192_pr --gfs_cyc 1 --idate 2020090600 --edate 2020090612 --configdir /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/parm/config --res 192 --comrot /lfs4/HFIP/hfv3gfs/Lin.Gan/ptmp/com --expdir /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir

SDATE = 2020-09-06 00:00:00
EDATE = 2020-09-06 12:00:00

EDITED:  /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/p3_c192_pr/config.base as per user input.
DEFAULT: /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/p3_c192_pr/config.base.default is for reference only.
Please verify and delete the default file before proceeding.

++ cd /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/p3_c192_pr
++ ln -s /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/ush/rocoto/setup_workflow_fcstonly.py

Modify config.base
  export PARTITION_BATCH="kjet"
  export DO_VRFY="NO"
  export REALTIME="NO"
  export FHMAX_GFS_00=168
  export FHMAX_GFS_06=168
  export FHMAX_GFS_12=168
  export FHMAX_GFS_18=168
  export FHOUT_GFS=6
  export DO_GLDAS="NO"
  export DO_METP="NO"         # Run METPLUS jobs - set METPLUS settings in config.metp
  export DO_VSDB="NO" 

(copy from /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/config.base)

[ =0 15:57:40 4981 Lin.Gan@fe3 ]
.../HFIP/hfv3gfs/Lin.Gan/expdir/p3_c192_pr> ./setup_workflow_fcstonly.py
sourcing config.getic
sourcing config.init
sourcing config.waveinit
sourcing config.waveprep
sourcing config.fcst
sourcing config.post
sourcing config.wavepostsbs
sourcing config.wavepostbndpnt
sourcing config.wavepostbndpntbll
sourcing config.wavepostpnt
sourcing config.wavegempak
sourcing config.waveawipsbulls
sourcing config.waveawipsgridded
sourcing config.wafs
sourcing config.wafsgrib2
sourcing config.wafsblending
sourcing config.wafsgcip
sourcing config.wafsgrib20p25
sourcing config.wafsblending0p25
sourcing config.postsnd
sourcing config.gempak
sourcing config.awips
sourcing config.vrfy
sourcing config.metp
sourcing config.arch

aerorahul commented on modulefiles/workflow_utils.jet:
```diff
@@ -0,0 +1,32 @@
+#%Module#####################################################
+## Workflow Utilities - hera
```
```suggestion
## Workflow Utilities - Jet
```

aerorahul commented on sorc/build_gfs_util.sh:
```diff
@@ -14,6 +14,12 @@ if [ ! -d "./mkgfsawps.fd" ]; then
 fi 
 
+if [ $target = jet ]; then
+   echo "  GFS_UTIL does not support JET "
```
```suggestion
   echo "JET does not support GFS_UTIL "
```

lgannoaa replied:
Done

aerorahul commented on sorc/build_gldas.sh:
```diff
@@ -4,6 +4,12 @@ set -eux
 cwd=`pwd`
 
+if [ $target = jet ]; then
+   echo "  GLDAS does not support JET "
```
```suggestion
   echo "JET does not support GLDAS"
```

lgannoaa replied:
Done

aerorahul commented on sorc/build_ww3prepost.sh:
```diff
@@ -10,6 +10,13 @@ finalexecdir=$( pwd -P )/../exec
 source ./machine-setup.sh > /dev/null 2>&1
+
+if [ $target = jet ]; then
+   echo "  WW3 does not support JET "
```
```suggestion
   echo "JET does not support WW3"
```

JessicaMeixner-NOAA replied:
Support for WW3 on Jet theoretically should work.  Not sure it's needed for this PR or in general just an FYI if its needed later.

lgannoaa replied:
WW3 is turned off for this PR.

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -244,6 +248,7 @@ def get_workflow(dict_configs, cdump='gdas'):
     do_gempak = base.get('DO_GEMPAK', 'NO').upper()
     do_awips = base.get('DO_AWIPS', 'NO').upper()
     do_wafs = base.get('WAFSF', 'NO').upper()
+    do_vrfy = base.get('DO_VRFY', 'NO').upper()
```
I don't think `DO_VRFY` is an entry in `config.base`.  By setting this to `NO` as default here, it is effectively turning off verification on all machines.

lgannoaa replied:
Modified to default to YES. User can set it to NO in config.base

aerorahul replied:
There is no entry in `config.base` for `DO_VRFY`.
Please add in the appropriate location in `config.base.emc.dyn` and `config.base.nco.static`.

lgannoaa replied:
Done.

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -683,7 +689,10 @@ def get_workflow(dict_configs, cdump='gdas'):
+    if do_vrfy in ['Y', 'YES']:
+        dep_dict = {'type':'task', 'name':f'{cdump}vrfy'}
+        deps.append(rocoto.add_dependency(dep_dict))
+    dep_dict = {'type':'streq', 'left':'&ARCHIVE_TO_HPSS;', 'right':'YES'}
```
Can you explain why `YES` is hard-wired here instead of `base.get('HPSSARCH', 'NO')` or `NO`.
`config.base` provides a toggle flag to archive to HPSS `HPSSARCH`.

lgannoaa replied:
Modified to dep_dict = {'type':'streq', 'left':'&ARCHIVE_TO_HPSS;', 'right':f'{hpssarch}'}

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,8 +204,11 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
-    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc'] and detectMachine() in ['HERA']:
```
You don't need the `and detectMachine() in ['HERA']` part here.

lgannoaa replied:
I did some modify to workflow_utils.py.

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -177,6 +177,10 @@ def get_resources(dict_configs, cdump='gdas'):
             strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
         if scheduler in ['slurm'] and task in ['getic', 'arch']:
             strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_SERVICE;">\n')
+        if scheduler in ['slurm'] and machine in ['JET'] and task in ['fcst']:
+            strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
+        if scheduler in ['slurm'] and machine in ['JET'] and task in ['post']:
+            strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
```
You don't need these lines if you change line 176 from:

```
if scheduler in ['slurm'] and machine in ['ORION'] and task not in ['getic', 'arch']:
```

to

```
if scheduler in ['slurm'] and machine in ['ORION',  'JET'] and task not in ['getic', 'arch']:
```

lgannoaa replied:
Will create different XML.
Recommend to keep the version to reproduce the control.

aerorahul replied:
Different XML's are fine.  I would like to see the XML and the differences (if any) resulting in the execution.

lgannoaa replied:
Modified as requested.

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,8 +204,11 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc'] and detectMachine() in ['HERA']:
+        task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
+    # Add PARTITION_SERVICE to all service jobs on JET (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc','fcst'] and detectMachine() in ['JET']:
```
Is `fcst` task run as part of the service job on JET? 
I don't think you need this line.

lgannoaa replied:
Modified fcst with PARTITION_BATCH

aerorahul reviewed (changes requested):
See code review comments.

aerorahul commented on sorc/build_gfs_util.sh:
```diff
@@ -14,6 +14,12 @@ if [ ! -d "./mkgfsawps.fd" ]; then
    exit
 fi 
 
+if [ $target = jet ]; then
```
@lgannoaa 
Instead of spreading the `if [ $target = jet ]; then` blocks across numerous build scripts, please add the following block on [line 42 of build_all.sh](https://github.com/NOAA-EMC/global-workflow/blob/8c777ad04bce0555bed9f6666b808515b0a3766d/sorc/build_all.sh#L42)

```
if [ $target = jet ]; then
  Build_gldas=false
  Build_gfs_util=false
  Build_ww3_prepost=false
fi
```

When eventually we enable Jet for these components, the changes are localized to this file and one does not have to go searching where to enable them.

lgannoaa replied:
Excellent suggestion, I modified it and removed the similar code from the individual scripts.

lgannoaa committed:
Minor restructure and logic modification.
Changes to be committed:
	modified:   modulefiles/workflow_utils.jet
	modified:   parm/config/config.base.emc.dyn
	modified:   parm/config/config.base.nco.static
	modified:   sorc/build_all.sh
	modified:   sorc/build_gfs_util.sh
	modified:   sorc/build_gldas.sh
	modified:   sorc/build_ww3prepost.sh
	modified:   ush/rocoto/setup_workflow_fcstonly.py
	modified:   ush/rocoto/workflow_utils.py

lgannoaa commented:
Please take a look again.

aerorahul commented on sorc/build_ww3prepost.sh:
```diff
@@ -10,6 +10,7 @@ finalexecdir=$( pwd -P )/../exec
 
 set +x
 source ./machine-setup.sh > /dev/null 2>&1
+
```
please remove

lgannoaa replied:
done

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -51,6 +51,7 @@ export DO_BUFRSND="NO"    # BUFR sounding products
 export DO_GEMPAK="NO"      # GEMPAK products
 export DO_AWIPS="NO"       # AWIPS products
 export WAFSF="NO"          # WAFS products
+export DO_VRFY="YES"        # VRFY step
```
please indent correctly.

lgannoaa replied:
done

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,8 +204,12 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
         task_dict['partition'] = '&PARTITION_BATCH;'
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
-    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc'] and detectMachine() in ['HERA','JET']:
+####    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
+        task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
+    # Add PARTITION_BATCH to all fcst jobs on JET (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['fcst'] and detectMachine() in ['JET']:
```
I think you dismissed my earlier comments without answering them.

lgannoaa replied:
I try to make change using your recommendation, but it created XML with many difference compare to my control. I would like to keep this version if possible.

aerorahul replied:
xml's are not committed to the repository and so it is irrelevant if they are different.  
They are different for different machines.

What matters is if the execution fails.

lgannoaa replied:
I made the change from your request. Testing it now.

aerorahul reviewed (changes requested):
progress.
Some comments need a response.

lgannoaa committed:
Clean up config.base.emc.dyn build_ww3prepost.sh and workflow_utils.py

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,8 +204,11 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
-    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc'] and detectMachine() in ['HERA','JET']:
+        task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
```
This is not just for Hera and Jet.  It is for all machines.

lgannoaa replied:
I will remove both HERA and JET here.

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,8 +204,11 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc'] and detectMachine() in ['HERA','JET']:
+        task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
+    # Add PARTITION_BATCH to all fcst jobs on JET (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['fcst'] and detectMachine() in ['JET']:
```
I think I asked this before, why is JET requiring special treatment for `fcst` and not for `post` or other jobs?

lgannoaa replied:
I believe it is because fcst job can run in different configuration to the best of using the JET partition. I did not recall you asked this question before.

lgannoaa replied:
Is the above two changes complete your request? Thanks

KateFriedman-NOAA replied:
I'd also like to understand why Jet is a special case here.

lgannoaa replied:
> I'd also like to understand why Jet is a special case here.

Answered.

lgannoaa committed:
Combine slurm platform in setup_workflow_fcstonly and workflow_utils logic

KateFriedman-NOAA commented on ush/rocoto/setup_expt_fcstonly.py:
```diff
@@ -207,6 +207,24 @@ def edit_baseconfig():
+      queue_service = 'service'
+      partition_batch = '$PARTITION_BATCH'
+      chgrp_rstprod = 'YES'
+      chgrp_cmd = 'chgrp rstprrod'
```
Remove extra r in 'rstprrod'. Should be 'rstprod'. Thanks!

lgannoaa replied:
done

KateFriedman-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -173,10 +173,14 @@ def get_resources(dict_configs, cdump='gdas'):
             strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
         if scheduler in ['slurm'] and task in ['getic', 'arch']:
             strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_SERVICE;">\n')
+#        if scheduler in ['slurm'] and machine in ['JET'] and task in ['fcst']:
+#            strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
+#        if scheduler in ['slurm'] and machine in ['JET'] and task in ['post']:
+#            strings.append(f'\t<!ENTITY PARTITION_{taskstr} "&PARTITION_BATCH;">\n')
```
Please remove if not needed. Thanks!

lgannoaa replied:
Ok

lgannoaa replied:
The fcst job failed during test with this change. I reverted the change and continue the test.

lgannoaa replied:
Found the issue; modified it and testing continue with fcst job in the queue.

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -683,7 +689,11 @@ def get_workflow(dict_configs, cdump='gdas'):
+    if do_vrfy in ['Y', 'YES']:
+        dep_dict = {'type':'task', 'name':f'{cdump}vrfy'}
+        deps.append(rocoto.add_dependency(dep_dict))
+#    dep_dict = {'type':'streq', 'left':'&ARCHIVE_TO_HPSS;', 'right':'YES'}
```
please remove commented line.

lgannoaa replied:
Done

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,9 +204,12 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
     if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION']:
         task_dict['partition'] = '&PARTITION_BATCH;'
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
```
```suggestion
    # Add PARTITION_SERVICE to all service jobs (SLURM)
```

lgannoaa replied:
Done

lgannoaa committed:
Minor fix on setup_expt_fcstonly

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,9 +204,12 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
     if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
         task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
+    # Add PARTITION_BATCH to all fcst jobs on JET (SLURM)
```
How is `post` job handled? Does it not be handled like the `fcst` job?
I think this should be for all (non-service) jobs on Jet, instead of just `fcst` jobs on Jet.

lgannoaa replied:
The line is removed as requested.

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,9 +204,12 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
     # Add PARTITION_BATCH to all non-service jobs on Orion (SLURM)
     if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION']:
         task_dict['partition'] = '&PARTITION_BATCH;'
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
     if get_scheduler(detectMachine()) in ['slurm'] and task in ['getic','arch','earc']:
         task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
+    # Add PARTITION_BATCH to all fcst jobs on JET (SLURM)
+    if get_scheduler(detectMachine()) in ['slurm'] and task in ['fcst'] and detectMachine() in ['JET']:
+        task_dict['partition'] = f'&PARTITION_{task.upper()}_{cdump.upper()};'
```
I think the logic here needs to be thought over.
This is supposed to be a generic method  to create tasks and metatasks (`create_wf_task`)
Over time if-blocks for machines and schedulers are added here and they should not.
These choices need to be made upstream of this method.

This comment is for a different issue, and will be addressed separately.

lgannoaa replied:
ok

KateFriedman-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -47,10 +47,11 @@ export NOSCRUB="@NOSCRUB@"
 export DO_GEMPAK="NO"      # GEMPAK products
 export DO_AWIPS="NO"       # AWIPS products
 export WAFSF="NO"          # WAFS products
+export DO_VRFY="YES"       # VRFY step
```
Why is a DO_VRFY toggle added to the system? The vrfy job runs more than just verification (VSDB), it also runs the GSI monitoring package, tracker, and fit2obs. If it's set to YES by default that is fine, most users will want it on...but this toggle isn't needed right now. We'll be breaking apart the vrfy job in the near future so this toggle will likely be replaced by other ones for the individual pieces. Seeing how you implemented this I am not opposed but it's not needed at this time. Do the tracker, monitoring package, and fit2obs work via the vrfy job on Jet?

lgannoaa replied:
DO_VRFY is a switch that can turn verification job off. The fcst only job does not run vrfy step. The METPlus are not currently supported in all platforms. We can add support when it is confirmed available later.

KateFriedman-NOAA replied:
Ok, if it's about turning off the METP jobs then the DO_METP variable in config.base should be set to NO when on Jet (it's currently YES by default everywhere). The METplus jobs are separate to the vrfy job. A new do_metp variable (similar to do_vrfy) in the setup scripts and configs should be added to turn off the gfsmetp jobs on Jet.

If the tracker and fit2obs pieces aren't yet working on Jet then yes we will want the vrfy job off and we can use the DO_VRFY switch you added. Please confirm those two pieces aren't yet working on Jet, thanks!

lgannoaa replied:
tracker and fit2obs are not part of this PR. Please open another issue as outlined in PR standard.

KateFriedman-NOAA commented on parm/config/config.base.nco.static:
```diff
@@ -50,6 +50,7 @@ export DO_BUFRSND="YES"    # BUFR sounding products
 export DO_GEMPAK="YES"     # GEMPAK products
 export DO_AWIPS="YES"      # AWIPS products
 export WAFSF="YES"         # WAFS products
+export DO_VRFY="NO"        # VRFY step
```
NCO doesn't run verification so this isn't needed. Can remove this line in config.base.nco.static.

lgannoaa replied:
NCO does not run it. That is why default to NO here.

KateFriedman-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -653,13 +654,14 @@ def get_workflow(dict_configs, cdump='gdas'):
-    task = wfu.create_wf_task('vrfy', cdump=cdump, envar=envars, dependency=dependencies)
-    tasks.append(task)
-    tasks.append('\n')
+    if do_vrfy in ['Y', 'YES']:
```
See my other comment about the need for this toggle.

lgannoaa replied:
Answered.

KateFriedman-NOAA commented on ush/rocoto/workflow_utils.py:
```diff
@@ -201,9 +204,12 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
     if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION']:
         task_dict['partition'] = '&PARTITION_BATCH;'
-    # Add PARTITION_SERVICE to all service jobs (SLURM)
+    # Add PARTITION_SERVICE to all service jobs on Hera (SLURM)
```
Why "on Hera"?

lgannoaa replied:
Has been removed.

KateFriedman-NOAA reviewed (changes requested):
Please address comments about DO_VRFY toggle and special Jet treatment in setup script. I will be running this branch on Jet and another platform to check it works. I will report any issues during testing in the PR comments, stay tuned for that. Thanks!

lgannoaa commented:
> Please address comments about DO_VRFY toggle and special Jet treatment in setup script. I will be running this branch on Jet and another platform to check it works. I will report any issues during testing in the PR comments, stay tuned for that. Thanks!

Please not to ask same question. It has been answered.

lgannoaa committed:
Remove comment line from setup_workflow_fcstonly and workflow_utils
Changes to be committed:
	modified:   ush/rocoto/setup_workflow_fcstonly.py
	modified:   ush/rocoto/workflow_utils.py

lgannoaa commented:
860d97e is the current version.

KateFriedman-NOAA commented on ush/rocoto/setup_expt_fcstonly.py:
```diff
@@ -207,6 +207,24 @@ def edit_baseconfig():
+      partition_batch = '$PARTITION_BATCH'
+      chgrp_rstprod = 'YES'
+      chgrp_cmd = 'chgrp rstprod'
+      exp_warm_start = '.false.'
```
Why was `exp_warm_start = '.false.'` added for Jet? If a user pulls down warm start ICs they should be able to run warm and set it as they need in config.base. This setting overrides the user's wishes when using the `--start` argument with the setup scripts right now. Please remove this line, thanks!

lgannoaa replied:
Ok

lgannoaa committed:
Remove exp_warm_start from setup_expt_fcstonly

aerorahul commented on ush/rocoto/setup_expt_fcstonly.py:
```diff
@@ -207,6 +207,23 @@ def edit_baseconfig():
+      account = 'hfv3gfs'
+      queue = 'batch'
+      queue_service = 'service'
+      partition_batch = '$PARTITION_BATCH'
```
`$PARTITION_BATCH` needs to be substituted with a value so a default needs to be provided.
```suggestion
      partition_batch = "${PARTITION_BATCH:-'xjet'}"
```

lgannoaa replied:
Why default everyone to use xjet on JET? I believe if a default is need, the python script should be modified to require config.base to have PARTITION set on JET as part of exception handling. May I keep this solution without change at this time?

aerorahul replied:
That is not true.
This line will add:
```
export PARTITION_BATCH=${PARTITION_BATCH:-'xjet'}
```
to `config.base` instead of:
```
export PARTITION_BATCH=$PARTITION_BATCH
```

If you don't provide a value, what is the value the script will pick?

For the user to use a specific partition, he user **has** to modify this line in `config.base` to provide a value or set it in the environment.

KateFriedman-NOAA replied:
PARTITION_BATCH in config.base REQUIRES a value so when PARTITION_BATCH goes through config.resources (as sourced by the setup_workflow*py scripts) it has to have a value for the Jet if-block in there. As it is right now the setup_workflow_fcstonly.py script fails:
```
-bash-4.2$ ./setup_workflow_fcstonly.py --expdir /lfs4/HFIP/hfv3gfs/Kate.Friedman/expdir/pr394ff
sourcing config.getic
sourcing config.init
sourcing config.waveinit
(standard_in) 1: syntax error
sourcing config.waveprep
(standard_in) 1: syntax error
sourcing config.fcst
(standard_in) 1: syntax error
(standard_in) 2: syntax error
(standard_in) 2: syntax error
sourcing config.post
sourcing config.wavepostsbs
(standard_in) 1: syntax error
sourcing config.wavepostbndpnt
(standard_in) 1: syntax error
sourcing config.wavepostbndpntbll
(standard_in) 1: syntax error
sourcing config.wavepostpnt
(standard_in) 1: syntax error
sourcing config.wavegempak
(standard_in) 1: syntax error
sourcing config.waveawipsbulls
(standard_in) 1: syntax error
sourcing config.waveawipsgridded
(standard_in) 1: syntax error
sourcing config.wafs
sourcing config.wafsgrib2
sourcing config.wafsblending
sourcing config.wafsgcip
sourcing config.wafsgrib20p25
sourcing config.wafsblending0p25
sourcing config.postsnd
sourcing config.gempak
sourcing config.awips
sourcing config.vrfy
sourcing config.metp
sourcing config.arch
Traceback (most recent call last):
  File "./setup_workflow_fcstonly.py", line 774, in <module>
    main()
  File "./setup_workflow_fcstonly.py", line 54, in main
    create_xml(dict_configs)
  File "./setup_workflow_fcstonly.py", line 744, in create_xml
    resources = get_resources(dict_configs, cdump=base['CDUMP'])
  File "./setup_workflow_fcstonly.py", line 171, in get_resources
    wtimestr, resstr, queuestr, memstr, natstr = wfu.get_resources(machine, cfg, task, reservation, cdump=cdump)
  File "/mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/ush/rocoto/workflow_utils.py", line 309, in get_resources
    nodes = np.int(np.ceil(np.float(tasks) / np.float(ppn)))
ValueError: could not convert string to float:
```

We need a default in PARTITION_BATCH in config.base....is there a desired default partition? xjet? kjet? other? Users can always change the partition in config.base before running setup_workflow*py to whichever partition they want to use other than whichever default we set.

Did you run the setup scripts from a cleanly cloned/built copy of this branch?

lgannoaa replied:
> As it is right now the setup_workflow_fcstonly.py script fails:
@KateFriedman-NOAA 
Please remove posted message for this session. The testing of setup_workflow_fcstonly.py should use python3.

lgannoaa replied:
> That is not true.
> This line will add:
> 
> ```
> export PARTITION_BATCH=${PARTITION_BATCH:-'xjet'}
> ```
> 
> to `config.base` instead of:
> 
> ```
> export PARTITION_BATCH=$PARTITION_BATCH
> ```
> 
> If you don't provide a value, what is the value the script will pick?
> 
> For the user to use a specific partition, he user **has** to modify this line in `config.base` to provide a value or set it in the environment.

I am making change base on your request and testing it now.

KateFriedman-NOAA replied:
I also get a syntax error with `export PARTITION_BATCH="${PARTITION_BATCH:-'xjet'}"` in config.base when setup_workflow_fcstonly.py sources it.

I spoke with @GeorgeVandenberghe-NOAA yesterday and he recommends we stick with just `xjet` for this initial pass at supporting Jet. The login nodes build for `xjet` and `kjet` by default right now so users who wish to use the other partitions will need to build for other partitions themselves for now. I ran my tests yesterday with `xjet` and the init/fcst/post jobs worked (getic and arch jobs used service queue). Users will want to turn off vrfy and metp jobs with DO_VRFY=NO and DO_METP=NO.

Please change the line in question in setup_expt_fcstonly.py to be:
```
partition_batch = 'xjet'
```

Thanks!

lgannoaa replied:
> ush/rocoto/rocoto_viewer.py

File rocoto_viewer.py was not supported in PR 370 that convert python2 to python3.
rocoto_viewer.py should get the same translation done before working on JET version.

KateFriedman-NOAA commented:
Successfully tested feature/jet-fcst-only branch on both WCOSS-Dell (Venus) and Hera:

WCOSS-Dell:

- HOMEgfs = /gpfs/dell2/emc/modeling/save/Kate.Friedman/git/global-workflow/feature-jet-fcst-only
- free-forecast (C192):
  - EXPDIR = /gpfs/dell2/emc/modeling/save/Kate.Friedman/expdir/pr394ff
  - COMROT = /gpfs/dell3/ptmp/Kate.Friedman/comrot/pr394ff
- cycled (C192C96L127):
  - EXPDIR = /gpfs/dell2/emc/modeling/save/Kate.Friedman/expdir/pr394cyc
  - COMROT = /gpfs/dell3/ptmp/Kate.Friedman/comrot/pr394cyc

Hera:

- HOMEgfs = /scratch1/NCEPDEV/global/Kate.Friedman/git/feature-jet-fcst-only
- free-forecast (C192):
  - EXPDIR = /scratch1/NCEPDEV/global/Kate.Friedman/expdir/pr394ff
  - COMROT = /scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/pr394ff
- cycled (C192C96L127):
  - EXPDIR = /scratch1/NCEPDEV/global/Kate.Friedman/expdir/pr394cyc
  - COMROT = /scratch1/NCEPDEV/stmp4/Kate.Friedman/comrot/pr394cyc

Need 2nd full cycle in cycled Hera test to complete but it's working fine otherwise.

Testing on Jet today now that it's back from maintenance. Will report back on how that testing goes.

lgannoaa committed:
Fix XML generation issue found during testing

lgannoaa committed:
Revert JET.env to get fcst job run

lgannoaa committed:
Fix launcher in JET.env and SENDCOM in config.post for fcst and post jobs to complete

KateFriedman-NOAA commented on ush/rocoto/setup_expt_fcstonly.py:
```diff
@@ -207,6 +207,23 @@ def edit_baseconfig():
+      account = 'hfv3gfs'
+      queue = 'batch'
+      queue_service = 'service'
+      partition_batch = "${PARTITION_BATCH:-'xjet'}"
```
Please change to:
`partition_batch = 'xjet'`

lgannoaa replied:
It can not be hard wired to xjet. JET can support many partitions.

KateFriedman-NOAA replied:
This sets the default that works best for users who don't know how to target another partitions and gets a partition correctly set in the resulting xml. Users can still change the partition in their respective config.base as needed (if they know what they are doing). Additional multi-partition support can be included in the full port work that will be done next. The support for different partitions will be explained by me when I announce this commit so don't worry about that. Please make the change I requested, thanks.

KateFriedman-NOAA commented on ush/rocoto/workflow_utils.py:
```diff
@@ -199,7 +202,7 @@ def create_wf_task(task, cdump='gdas', cycledef=None, envar=None, dependency=Non
 
     # Add PARTITION_BATCH to all non-service jobs on Orion (SLURM)
-    if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION']:
+    if get_scheduler(detectMachine()) in ['slurm']:
```
This line should be:
```
if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION','JET']:
```
...otherwise the needed partition entity isn't added to each task in the xml and the jobs end up on random partitions (setting partition_batch in config.base gets ignored without this).

lgannoaa replied:
In my test, the xml file contains required PARTITION_BATCH. Please see /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/test3_c192_pr/test3_c192_pr.xml

@KateFriedman-NOAA 
Please include error message or location of the file you are reporting within the message going forward.

KateFriedman-NOAA replied:
So yes PARTITION_BATCH is added to the top of the generated xml but the needed <partition> setting in each task section is missing without the change I requested. See this diff between two xml's I tested on Jet:

```
-bash-4.2$ pwd
/lfs4/HFIP/hfv3gfs/Kate.Friedman/expdir/pr394ffb
-bash-4.2$ diff ../pr394ff/pr394ff.xml pr394ffb.xml
...
257a258
>       <partition>&PARTITION_BATCH;</partition>
294a296
>       <partition>&PARTITION_BATCH;</partition>
333a336
>               <partition>&PARTITION_BATCH;</partition>
366a370
>       <partition>&PARTITION_BATCH;</partition>
398a403
>               <partition>&PARTITION_BATCH;</partition>
```

The pr394ff run (without `<partition>&PARTITION_BATCH;</partition>` in each batch/non-service task) submitted jobs to whichever partition was the quietist (not what we want). My pr394ffb run (with the `<partition>` settings) correctly submitted all batch jobs to PARTITION_BATCH set at the top of the xml (xjet in this case). Without the `<partition>` setting in each task in the xml the user requested partition is completely ignored.

Also, by removing the Orion part of that line you are breaking Orion support. Please make the change I requested, thanks!

aerorahul replied:
@lgannoaa 
@KateFriedman-NOAA is testing this branch on all other platforms to ensure the changes in this PR are not breaking them.
Would you like to test your PR on other platforms to make sure they don't break there?

lgannoaa replied:
The test from pr394ffb and mine (/lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/test2_c192_pr/test2_c192_pr.xml) are the same on job partition syntax.
The difference is pr394ffb run unnecessary steps and is using xjet and test2_c192_pr.xml is kjet. 
I recall the action of remove "and detectMachine() in ['ORION','JET']:" was requested by @aerorahul originally.

KateFriedman-NOAA replied:
@lgannoaa Please make the following two changes and we can be done with this discussion about partitions:

1) set partition default to 'xjet' in setup_expt_fcstonly.py (`partition_batch = 'xjet'`), users can change to kjet if they wish at setup time
2) change line in ush/rocoto/workflow_utils.py:
`if get_scheduler(detectMachine()) in ['slurm'] and detectMachine() in ['ORION','JET']:`

@aerorahul has given a thumbs up to both of these, see thumbs up emoji reactions to my requests above

WalterKolczynski-NOAA replied:
Wouldn't Lin's change do the opposite and add it to every machine that uses slurm, so it (might) break Hera rather than Orion?

That said, Hera does have a partition setting as well.

KateFriedman-NOAA commented on modulefiles/module_base.jet:
```diff
@@ -1,32 +1,24 @@
-     puts stderr "The prerequisites for compiling or running FV3 on Jet. "
-}
+# Load hpc-stack
+module use /lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/modulefiles/stack
```
This module_base for Jet needs a full overhaul, it does not work on Jet. Please grab this copy from my test clone on Jet:

/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet

I ran successfully with that updated module_base.jet yesterday. Thanks!

lgannoaa replied:
> This module_base for Jet needs a full overhaul, it does not work on Jet.

The current module_base.jet works. Again, please include error message to justify why it does not work for your test.

KateFriedman-NOAA replied:
@lgannoaa See this fcst job log on Jet:
/lfs4/HFIP/hfv3gfs/Kate.Friedman/ptmp/comrot/pr394ff/logs/2021073100/gfsfcst.log.1

The job failed because it didn't know what setpdy was because the prod_util module wasn't loaded because it's wrong in module_base.jet (prod_util/v1.0.18). Upon further inspection of module_base.jet I saw that grib_util was also wrong (not the stack one) and the other modules hadn't been updated to match versions in the stack being used on other platforms. The module_base.jet I pointed you to (which I updated and tested yesterday) was based on module_base.hera with needed updates for Jet (e.g. stack path). Please pull in the module_base.jet version I pointed you to. Thanks!

lgannoaa replied:
/lfs4/HFIP/hfv3gfs/Lin.Gan/ptmp/com/test2_c192_pr/logs/2020090600/gfsfcst.log line 1174 shown my test successfully run with setpdy.sh executed.
Please redo your test using 97327d. Thanks

KateFriedman-NOAA replied:
I updated my clone of your PR branch to the newer hash (97327d) and reran a gfsfcst job with module_base.jet as you have set it and it again failed with the same error. See log here:

/lfs4/HFIP/hfv3gfs/Kate.Friedman/ptmp/comrot/pr394ff/logs/2021073100/gfsfcst.log.0

I then only changed the prod_util module to be the hpc-stack one, reran the gfsfcst again and it's now running successfully:

/lfs4/HFIP/hfv3gfs/Kate.Friedman/ptmp/comrot/pr394ff/logs/2021073100/gfsfcst.log

I am running with a relatively clean account that only loads the stack in my environment. Because of that and the fact that you aren't specifying where that `prod_util/v1.0.18` module can be found the job can't find it at runtime. I'm guessing you have a module use in your Jet .bashrc that helps? I managed to find the `prod_util/v1.0.18` module on Jet and I see that prod_util version is coming from an old copy of NCEPLIBS (/lfs4/HFIP/hfv3gfs/nwprod/NCEPLIBS) which we should be moving away from. So, regardless of all that, you should be using the hpc-stack versions of prod_util and grib_util (and other modules that are available via hpc-stack). You should also be including the other modules that are missing in module_base.jet. So, again, please grab the copy I modified and successfully ran with and add it to your branch. I renamed it before testing your version again so please grab this copy to overwrite your copy of module_base.jet:

/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet_GOOD

lgannoaa replied:
1. module-setup.sh.inc has module purge on Jet

2. Please explain why your testing module_base.jet is different if you are using my PR branch (97327d) 
diff /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/modulefiles/module_base.jet
18,19c18
< #module load prod_util/v1.0.18
< module load prod_util/1.2.2
---
> module load prod_util/v1.0.18

**I am unable to reproduce the issue you reported.**
---
[ =0 18:47:04 5792 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> module purge
[ =0 18:47:12 5793 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> which setpdy.sh
/usr/bin/which: no setpdy.sh in ...
module use /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles
module load module_base.jet
which setpdy.sh
**/lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/intel-18.0.5.274/prod_util/1.2.2/bin/setpdy.sh**
---

lgannoaa replied:
1. module-setup.sh.inc has module purge on Jet
2. Please explain why your testing module_base.jet is different if you are using my PR branch (97327d) 
diff /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/modulefiles/module_base.jet
18,19c18
< #module load prod_util/v1.0.18
< module load prod_util/1.2.2
---
> module load prod_util/v1.0.18

I am unable to reproduce the issue you reported.
---
[ =0 18:47:04 5792 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> module purge
[ =0 18:47:12 5793 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> which setpdy.sh
/usr/bin/which: no setpdy.sh in ...
module use /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles
module load module_base.jet
which setpdy.sh
/lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/intel-18.0.5.274/prod_util/1.2.2/bin/setpdy.sh
---

lgannoaa replied:
1. module-setup.sh.inc has module purge on Jet
2. Please explain why your testing module_base.jet is different if you are using my PR branch (97327d) 
diff /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/modulefiles/module_base.jet
18,19c18
< #module load prod_util/v1.0.18
< module load prod_util/1.2.2
---
> module load prod_util/v1.0.18

I am unable to reproduce the issue you reported.

/lfs4/HFIP/hfv3gfs/Lin.Gan> module purge
[ =0 18:47:12 5793 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> which setpdy.sh
/usr/bin/which: no setpdy.sh in ...
module use /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles
module load module_base.jet
which setpdy.sh
/lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/intel-18.0.5.274/prod_util/1.2.2/bin/setpdy.sh

KateFriedman-NOAA replied:
> Please explain why your testing module_base.jet is different if you are using my PR branch (97327d)
diff /mnt/lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/modulefiles/module_base.jet
18,19c18
< #module load prod_util/v1.0.18
< module load prod_util/1.2.2

I already explained that, please reread my last comment. You're seeing the one module change I made to get it working (gfsfcst.log) after it failed again with the modules as you have set (gfsfcst.log.0).

What do you have in your .bashrc? Please show what module and PATH settings you have in your .bashrc (e.g. PATH, module use, and module load statements). I am seeing MANY more references to the old NCEPLIBS area (/lfs4/HFIP/hfv3gfs/nwprod/NCEPLIBS) in your gfsfcst.log than either of mine which means your account is getting that path from somewhere.

But, again, this is irrelevant, you should be using hpc-stack modules and have a complete module_base.jet. Your module_base.jet is wrong. Please replace with the modified one I provided.

lgannoaa replied:
Please keep environment were issue was reported. So other can reproduce the issue as needed.

/lfs4/HFIP/hfv3gfs/Lin.Gan> module purge
[ =0 18:46:38 5788 Lin.Gan@fe8 ]
/lfs4/HFIP/hfv3gfs/Lin.Gan> which setpdy.sh
/usr/bin/which: no setpdy.sh in
module use /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr/modulefiles
/lfs4/HFIP/hfv3gfs/Lin.Gan> module load module_base.jet
/lfs4/HFIP/hfv3gfs/Lin.Gan> which setpdy.sh
/lfs4/HFIP/hfv3gfs/nwprod/NCEPLIBS/src/prod_util_v1.0.18/ush/setpdy.sh
bashrc is not an issue. module prod_util_v1.0.18 loaded provided access to setpdy.sh.

Change made as requested to use hpc-stack.

GeorgeVandenberghe-NOAA commented:
I have used hpc-stack based modules for UPP on jet.  These work on all
partitions.   This is NOT sufficient reason to try to support all
partitions; there are too many other hassles
bash-4.2$ cat jet
*#%Module#######################################################################
Wen Meng 01/2021, Set up config. with the hpc-stack
NCEPLIBS.##############################################################################proc
ModulesHelp { } {puts stderr "Loads modules required for building
upp"}module-whatis "Loads UPP prerequisites on Jet"module load
cmake/3.16.1module use
/lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/modulefiles/stackmodule load
hpc/1.1.0module load hpc-intel/18.0.5.274module load
hpc-impi/2018.4.274module load jasper/2.0.22module load zlib/1.2.11module
load png/1.6.35module load hdf5/1.10.6module load netcdf/4.7.4module load
bacio/2.4.1module load crtm/2.3.0module load g2/3.4.1module load
g2tmpl/1.10.0module load ip/3.3.3module load nemsio/2.5.2module load
sfcio/1.4.1module load sigio/2.3.2module load sp/2.3.3module load
w3nco/2.4.1module load w3emc/2.7.3module load wrf_io/1.1.1*



On Fri, Aug 6, 2021 at 9:56 AM Kate Friedman ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In modulefiles/module_base.jet
> <https://github.com/NOAA-EMC/global-workflow/pull/394#discussion_r684256217>
> :
>
> >  ##
>
> -proc ModulesHelp {} {
> -     puts stderr "The prerequisites for compiling or running FV3 on Jet. "
> -}
> +# Load hpc-stack
> +module use /lfs4/HFIP/hfv3gfs/nwprod/hpc-stack/libs/modulefiles/stack
>
> This module_base for Jet needs a full overhaul, it does not work on Jet.
> Please grab this copy from my test clone on Jet:
>
>
> /lfs4/HFIP/hfv3gfs/Kate.Friedman/git/feature-jet-fcst-only/modulefiles/module_base.jet
>
> I ran successfully with that updated module_base.jet yesterday. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/394#pullrequestreview-724385322>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FSDMT6BATOKVPHKWXLT3PS2DANCNFSM5BPHANXA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

lgannoaa committed:
Turn GOESF to NO in matching current EMC parallel default

KateFriedman-NOAA commented:
To get the viewer working on Jet please make the following change to ush/rocoto/rocoto_viewer.py, thanks!
```
diff --git a/ush/rocoto/rocoto_viewer.py b/ush/rocoto/rocoto_viewer.py
index 77c59ca0..33584e40 100755
--- a/ush/rocoto/rocoto_viewer.py
+++ b/ush/rocoto/rocoto_viewer.py
@@ -193,12 +193,13 @@ def load_produtil_pythonpath():

     PRODUTIL = collections.defaultdict(list)
     PRODUTIL['hera'] = '/scratch1/NCEPDEV/global/glopara/svn/nceplibs/produtil/trunk/ush'
+    PRODUTIL['jet'] = '/lfs4/HFIP/hfv3gfs/glopara/svn/nceplibs/produtil/trunk/ush'
     PRODUTIL['orion'] = '/work/noaa/global/glopara/svn/nceplibs/produtil/trunk/ush'
     PRODUTIL['luna']  = '/gpfs/hps3/emc/global/noscrub/emc.glopara/svn/nceplibs/produtil/trunk/ush'
     PRODUTIL['surge']  = '/gpfs/hps3/emc/global/noscrub/emc.glopara/svn/nceplibs/produtil/trunk/ush'
     PRODUTIL['mars']  = '/gpfs/dell2/emc/modeling/noscrub/emc.glopara/svn/nceplibs/produtil/trunk/ush'
     PRODUTIL['venus']  = '/gpfs/dell2/emc/modeling/noscrub/emc.glopara/svn/nceplibs/produtil/trunk/ush'
-    try_clusters = ('hera','orion','luna','surge','mars','venus')
+    try_clusters = ('hera','jet','orion','luna','surge','mars','venus')

     for cluster in try_clusters:
         sys.path.append(PRODUTIL[cluster])
```

lgannoaa commented:
Minor update made on this version 97327d3.
All GFS fcst only job for a cycle completed during test on 8/5-8/6.
HOMEgfs: /mnt/lfs4/HFIP/hfv3gfs/Lin.Gan/PR/global-workflow-pr
expdir: /lfs4/HFIP/hfv3gfs/Lin.Gan/expdir/test2_c192_pr
log: /lfs4/HFIP/hfv3gfs/Lin.Gan/ptmp/com/test2_c192_pr/logs/2020090600
com: /lfs4/HFIP/hfv3gfs/Lin.Gan/ptmp/com/test2_c192_pr/gfs.20200906/00/atmos/

May I proceed to the regular PR?

aerorahul commented:
Yes @lgannoaa 
You may mark it as "Ready for review".
The review continues.

lgannoaa commented:
> To get the viewer working on Jet please make the following change to ush/rocoto/rocoto_viewer.py

File rocoto_viewer.py was not supported in PR 370 that convert python2 to python3.
rocoto_viewer.py should get the same translation done before working on JET version.

WalterKolczynski-NOAA commented:
Please try the rocoto_viewer from this pending CROW PR: https://github.com/NOAA-EMC/CROW/pull/16

I haven't tested it on Jet, but it updates to python 3 and removes the produtil script dependency, so it shouldn't have hard-coded machine dependencies anymore. We're still figuring out where the best home for the script actually is, since we have different copies in different places, and I'm not sure any of them are where it actually belongs.

lgannoaa commented:
> Please try the rocoto_viewer from this pending CROW PR
Can we proceed rocoto_viewer.py update in a separate PR? As I reported, It should go through the same translation work as other python on PR 370. Currently this file in global-workflow develop version does not consistent on syntax with other python3 scripts.

aerorahul commented:
> > Please try the rocoto_viewer from this pending CROW PR
> > Can we proceed rocoto_viewer.py update in a separate PR? As I reported, It should go through the same translation work as other python on PR 370. Currently this file in global-workflow develop version does not consistent on syntax with other python3 scripts.

@lgannoaa
We can leave the rocoto viewer update out of this review.

WalterKolczynski-NOAA commented on env/JET.env:
```diff
@@ -103,12 +104,8 @@ elif [ $step = "fcst" ]; then
-    else
-      export APRUN_FV3="$launcher ${npe_fcst_gfs:-$PBS_NP}"
-    fi
+    export APRUN_FV3="$launcher"
```
Is this going to run into the same issue we just saw in the sprint branch by not specifying the number of tasks?

KateFriedman-NOAA replied:
Possibly, I'd need to see the issue in the sprint branch more to understand...but this works on Hera and Jet without specifying a task number. Adding a npe value behind the $launcher setting (as it was in JET.env) causes an error. The contents of JET.env were very outdated so this brings APRUN_FV3 in line with changes in the other env files on SLURM machines.

WalterKolczynski-NOAA replied:
On Orion, we were over-requesting CPU resources relative to the layout (240 tasks vs 224 needed) and FV3 was getting the 240 from srun and complaining. This might be a recent change to UFS that will have to be addressed everywhere if we are over-requesting elsewhere.

KateFriedman-NOAA replied:
Gotcha, we can address that in a separate issue.

KateFriedman-NOAA commented:
> Wouldn't Lin's change do the opposite and add it to every machine that uses slurm, so it (might) break Hera rather than Orion? That said, Hera does have a partition setting as well.

@WalterKolczynski-NOAA My test of this PR branch on Hera had the partition set for the service jobs but not the batch jobs (because of how Lin changed that line) so yes SLURM on Hera also supports partitions but isn't required to run there like it's required on Orion/Jet. I would need to redo my test on Hera (with `<!ENTITY PARTITION_BATCH "hera">`) and do a test on Orion to confirm it works but this would lead to further changes to adjust things for Hera/Orion and touches on a larger discussion about how we're handling partitions. So I'd like to just wrap this up with minimal changes for Jet and then rework the partition stuff in a separate targeted PR. This PR is meant to just be about Jet fcst-only...so simply adding `'JET'` to the line in question is fast/cheap for now. Appreciate your inputs as always though, thanks!

lgannoaa committed:
Restore test for Orion and Jet back to workflow_utils

lgannoaa committed:
Wire partition_batch to xjet in setup_expt_fcstonly as requested

lgannoaa commented:
Change made as requested. Please use 687c5d going forward.

lgannoaa committed:
Using module load prod_util/1.2.2 in module_base.jet
 Changes to be committed:
	modified:   modulefiles/module_base.jet

lgannoaa commented:
Change made as requested. Please use 7c3777 going forward.

arunchawla-NOAA commented:
There have been multiple back and forths on this PR. Where are we now ? Have all the issues that were raised been addressed ? Are there outstanding issues still left? Lin please make sure that the Jet Module is using the appropriate hpc-stack for Jet. It is time to finish this as there will be many more updates that will need to be made. Of course the PR has to satisfy the review process and all code management issues raised need to be addressed (NOTE: If any issue is being moved to a future PR then make sure that the appropriate issues are created)

lgannoaa commented:
> Lin please make sure that the Jet Module is using the appropriate hpc-stack for Jet
I believe this request has been done with 7c3777.
Moving PR to ready for review.

aerorahul reviewed (approved):
Changes look good.
Will leave final approval and merge to the code managers pending their testing.

WalterKolczynski-NOAA commented on modulefiles/module_base.jet:
```diff
@@ -1,32 +1,24 @@
 module load gempak/7.4.2
-
+module load prod_util/1.2.2
+module load grib_util/v1.1.1
```
Some of these module versions are still not the correct ones located in hpc-stack. Please compare this modulefile with the one from Hera that already uses hpc-stack and ensure the module versions are the same for all that are available under hpc-stack.

lgannoaa replied:
I compared each one of the module. Found the following module on Hera is not in hpc-stack:
hpss/hpss
nco/4.9.1
cdo/1.9.5
gempak/7.4.2
intel/18.0.5.274
impi/2018.4.274

I made update to include:
grib_util/1.2.2
g2tmpl/1.10.0

WalterKolczynski-NOAA replied:
The g2tmpl version should be the same as the other machines (1.9.1).

lgannoaa replied:
> The g2tmpl version should be the same as the other machines (1.9.1).

Hera used g2tmpl/1.10.0 because module upp/10.0.6:

module load upp/10.0.6
The following have been reloaded with a version change:
  1) g2/3.4.1 => g2/3.4.3     2) g2tmpl/1.9.1 => g2tmpl/1.10.0     3) w3emc/2.7.3 => w3emc/2.9.0


Therefore, I used g2tmpl/1.10.0 on JET

WalterKolczynski-NOAA replied:
Okay, we'll have to update the other modulefiles separately then.

lgannoaa committed:
Add grib_util/1.2.2 and g2tmpl/1.10.0 into module_base.jet
Changes to be committed:
	modified:   modulefiles/module_base.jet

KateFriedman-NOAA reviewed (approved):
Requested changes have been sufficiently addressed. I will submit PR shortly.

KateFriedman-NOAA commented:
This PR partially completes work for issue #357
